libs/architectures/mlpe/architectures/flows/maf.py

```diff
@@ -0,0 +1,88 @@
+from dataclasses import dataclass
+from typing import Callable, Optional, Tuple
+
+import torch
+from mlpe.architectures.embeddings.convolution import NaiveEmbedding
+from mlpe.architectures.flows.flow import NormalizingFlow
+from nflows.distributions import StandardNormal
+from nflows.flows import Flow
+from nflows.transforms import CompositeTransform, RandomPermutation
+from nflows.transforms.autoregressive import (
+    MaskedAffineAutoregressiveTransform,
+)
+
+
+@dataclass
+class MaskedAutoRegressiveFlow(NormalizingFlow):
+    shape: Tuple[int, int]
+    num_transforms: int = 5
+    num_embedding_dim: Optional[int] = None
+    hidden_features: int = 50
+    num_blocks: int = 2
+    activation: Callable = torch.tanh
+    use_batch_norm: bool = False
+    use_residual_blocks: bool = True
+
+    def __post_init__(self):
+        self._flow = None
+        self.param_dim = self.shape[0]
+        if not self.num_embedding_dim:
+            self.embedding_net = None
+            self.context_dim = self.shape[1]
+        else:
+            self.embedding_net = NaiveEmbedding(
+                self.shape[1],
+                self.num_embedding_dim,
+                activation=self.activation,
+                hidden_layer_size=100,
+                num_hidden_layers=2,  # FIXME: don't hardcode
+            )
+            # infer context dim by passing ones
+            _sample_input = torch.ones(self.shape)[1]
+            self.context_dim = self.embedding_net(_sample_input).numel()
+
+        super().__init__(
+            self.param_dim,
+            self.context_dim,
+            num_flow_steps=self.num_transforms,
+            embedding_net=self.embedding_net,
+        )
+
+    def transform_block(self, idx: int):
+        """Returns the single block of the MAF"""
+        single_block = [
+            MaskedAffineAutoregressiveTransform(
+                features=self.param_dim,
+                hidden_features=self.hidden_features,
+                context_features=self.context_dim,
+                num_blocks=self.num_blocks,
+                activation=self.activation,
+                use_batch_norm=self.use_batch_norm,
+                use_residual_blocks=self.use_residual_blocks,
+            ),
+            RandomPermutation(features=self.param_dim),
+        ]
+        return single_block
+
+    def linear_block(self):
```

Comment:
Given this, it might make sense to remove the `@abstractmethod` from the base class. Users who wan't to create a flow with a linear method should then add that in the `transform_block` method.

Reply:
done